File: packages/ispf-core/src/main/java/com/ispf/core/ref/PlatformRef.java
package com.ispf.core.ref;

import com.ispf.core.binding.BindingVariableRef;

import java.util.Objects;

/**
 * Canonical address of a variable, function, event, or historian tag on the object tree.
 *
 * <p>Text form (slash grammar): see ADR-0043.
 */
public record PlatformRef(
        String object,
        PlatformRefKind kind,
        String name,
        String field
) {

    public static final String CURRENT_OBJECT = "@";
    public static final String DEFAULT_FIELD = "value";

    public PlatformRef {
        if (object == null || object.isBlank()) {
            throw new IllegalArgumentException("object is required");
        }
        if (kind == null) {
            kind = PlatformRefKind.VARIABLE;
        }
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("name is required");
        }
        if (kind == PlatformRefKind.VARIABLE) {
            if (field == null || field.isBlank()) {
                field = DEFAULT_FIELD;
            }
        } else {
            field = null;
        }
    }

    public static PlatformRef variable(String object, String variableName) {
        return variable(object, variableName, DEFAULT_FIELD);
    }

    public static PlatformRef variable(String object, String variableName, String field) {
        return new PlatformRef(object, PlatformRefKind.VARIABLE, variableName, field);
    }

    public static PlatformRef function(String object, String functionName) {
        return new PlatformRef(object, PlatformRefKind.FUNCTION, functionName, null);
    }

    public static PlatformRef event(String object, String eventName) {
        return new PlatformRef(object, PlatformRefKind.EVENT, eventName, null);
    }

    public static PlatformRef tag(String object, String ruleId) {
        return new PlatformRef(object, PlatformRefKind.TAG, ruleId, null);
    }

    public static PlatformRef currentVariable(String variableName) {
        return variable(CURRENT_OBJECT, variableName);
    }

    public static PlatformRef currentVariable(String variableName, String field) {
        return variable(CURRENT_OBJECT, variableName, field);
    }

    public boolean isCurrentObject() {
        return CURRENT_OBJECT.equals(object);
    }

    public boolean isVariable() {
        return kind == PlatformRefKind.VARIABLE;
    }

    public boolean isFunction() {
        return kind == PlatformRefKind.FUNCTION;
    }

    public boolean isEvent() {
        return kind == PlatformRefKind.EVENT;
    }

    public boolean isTag() {
        return kind == PlatformRefKind.TAG;
    }

    /**
     * Resolves {@code @} to the rule-owner path for dependency indexing.
     */
    public PlatformRef resolveObject(String ruleObjectPath) {
        if (!isCurrentObject()) {
            return this;
        }
        if (ruleObjectPath == null || ruleObjectPath.isBlank()) {
            throw new IllegalArgumentException("ruleObjectPath is required to resolve @");
        }
        return switch (kind) {
            case VARIABLE -> variable(ruleObjectPath, name, field);
            case FUNCTION -> function(ruleObjectPath, name);
            case EVENT -> event(ruleObjectPath, name);
            case TAG -> tag(ruleObjectPath, name);
        };
    }

    public BindingVariableRef toVariableActivatorRef(String ruleObjectPath) {
        if (kind != PlatformRefKind.VARIABLE) {
            throw new IllegalStateException("Not a variable ref: " + this);
        }
        PlatformRef resolved = resolveObject(ruleObjectPath);
        if (resolved.object().equals(ruleObjectPath)) {
            return BindingVariableRef.local(resolved.name());
        }
        return BindingVariableRef.remote(resolved.object(), resolved.name());
    }

    public String eventKey(String ruleObjectPath) {
        PlatformRef resolved = resolveObject(ruleObjectPath);
        return resolved.object() + "|" + resolved.name();
    }

    @Override
    public String toString() {
        return PlatformRefFormatter.format(this);
    }

    public boolean structurallyEquals(PlatformRef other) {
        if (other == null) {
            return false;
        }
        return Objects.equals(object, other.object)
                && kind == other.kind
                && Objects.equals(name, other.name)
                && Objects.equals(field, other.field);
    }
}

File: apps/web-console/src/utils/platformRef.ts
/**
 * PlatformRef slash grammar (ADR-0043) — TS mirror of com.ispf.core.ref.PlatformRefParser.
 */

export type PlatformRefKind = "variable" | "function" | "event" | "tag";

export interface PlatformRef {
  object: string;
  kind: PlatformRefKind;
  name: string;
  field?: string;
}

const IDENT = /^[A-Za-z_][A-Za-z0-9_]*$/;
const SLASH_FN = /^(@|[A-Za-z_][A-Za-z0-9_.-]*)\/fn\/([A-Za-z_][A-Za-z0-9_-]+)$/;
const SLASH_EVT = /^(@|[A-Za-z_][A-Za-z0-9_.-]*)\/evt\/([A-Za-z_][A-Za-z0-9_-]+)$/;
const SLASH_TAG = /^(@|[A-Za-z_][A-Za-z0-9_.-]*)\/tag\/([A-Za-z_][A-Za-z0-9_-]+)$/;
const SLASH_VAR = /^(@|[A-Za-z_][A-Za-z0-9_.-]*)\/([A-Za-z_][A-Za-z0-9_-]+)(?:\/([A-Za-z_][A-Za-z0-9_]+))?$/;

export function parsePlatformRef(raw: string): PlatformRef | null {
  const trimmed = raw?.trim();
  if (!trimmed) return null;

  let m = trimmed.match(SLASH_FN);
  if (m) return { object: m[1], kind: "function", name: m[2] };

  m = trimmed.match(SLASH_EVT);
  if (m) return { object: m[1], kind: "event", name: m[2] };

  m = trimmed.match(SLASH_TAG);
  if (m) return { object: m[1], kind: "tag", name: m[2] };

  m = trimmed.match(SLASH_VAR);
  if (m) {
    return {
      object: m[1],
      kind: "variable",
      name: m[2],
      field: m[3] ?? "value",
    };
  }

  if (IDENT.test(trimmed)) {
    return { object: "@", kind: "variable", name: trimmed, field: "value" };
  }

  return null;
}

export function formatPlatformRef(ref: PlatformRef): string {
  switch (ref.kind) {
    case "function":
      return `${ref.object}/fn/${ref.name}`;
    case "event":
      return `${ref.object}/evt/${ref.name}`;
    case "tag":
      return `${ref.object}/tag/${ref.name}`;
    case "variable":
    default:
      if (ref.field && ref.field !== "value") {
        return `${ref.object}/${ref.name}/${ref.field}`;
      }
      return `${ref.object}/${ref.name}`;
  }
}

export function refFromFields(
  objectPath: string | undefined,
  name: string | undefined,
  field?: string,
  kind: PlatformRefKind = "variable"
): string | undefined {
  if (!name?.trim()) return undefined;
  const object = !objectPath || objectPath === "self" ? "@" : objectPath.trim();
  if (kind === "function") return formatPlatformRef({ object, kind, name: name.trim() });
  if (kind === "event") return formatPlatformRef({ object, kind, name: name.trim() });
  return formatPlatformRef({
    object,
    kind: "variable",
    name: name.trim(),
    field: field?.trim() || "value",
  });
}

export function fieldsFromRef(refString: string | undefined): {
  objectPath?: string;
  name?: string;
  field?: string;
  kind: PlatformRefKind;
} {
  if (!refString?.trim()) return { kind: "variable" };
  const parsed = parsePlatformRef(refString);
  if (!parsed) return { kind: "variable" };
  return {
    objectPath: parsed.object === "@" ? "self" : parsed.object,
    name: parsed.name,
    field: parsed.field,
    kind: parsed.kind,
  };
}
